src/compiler/tsc.ts

```diff
@@ -146,131 +173,176 @@ module ts {
-                var filename = getCanonicalName(f.filename);
-                if (hasProperty(watchers, filename)) {
-                    watchers[filename].close();
+        performCompilation();
+
+        // Invoked to perform initial compilation or re-compilation in watch mode
+        function performCompilation() {
+
+            if (!cachedSourceFiles) {
+                if (configFilename) {
+                    var configObject = readConfigFile(configFilename);
+                    if (!configObject) {
+                        reportDiagnostic(createCompilerDiagnostic(Diagnostics.Unable_to_open_file_0, configFilename));
+                        return sys.exit(EmitReturnStatus.CompilerOptionsErrors);
+                    }
+                    var configParseResult = parseConfigFile(configObject, getDirectoryPath(configFilename));
+                    if (configParseResult.errors.length > 0) {
+                        reportDiagnostics(configParseResult.errors);
+                        return sys.exit(EmitReturnStatus.CompilerOptionsErrors);
+                    }
+                    rootFilenames = configParseResult.filenames;
+                    compilerOptions = extend(commandLine.options, configParseResult.options);
                 }
-            });
+                else {
+                    rootFilenames = commandLine.filenames;
+                    compilerOptions = commandLine.options;
+                }
+                compilerHost = createCompilerHost(compilerOptions);
+                hostGetSourceFile = compilerHost.getSourceFile;
+                compilerHost.getSourceFile = getSourceFile;
+            }
 
-            watchers = {};
-        }
+            var compileResult = compile(rootFilenames, compilerOptions, compilerHost);
 
-        // Fired off whenever a file is changed.
-        function fileUpdated(filename: string) {
-            var firstNotification = isEmpty(updatedFiles);
-            updatedFiles[getCanonicalName(filename)] = true;
+            if (!commandLine.options.watch) {
+                return sys.exit(compileResult.exitStatus);
+            }
 
-            // Only start this off when the first file change comes in,
-            // so that we can batch up all further changes.
-            if (firstNotification) {
-                setTimeout(() => {
-                    var changedFiles = updatedFiles;
-                    updatedFiles = {};
+            updateSourceFileCache(compileResult.program.getSourceFiles());
+            reportDiagnostic(createCompilerDiagnostic(Diagnostics.Compilation_complete_Watching_for_file_changes));
+        }
 
-                    recompile(changedFiles);
-                }, 250);
+        function getSourceFile(filename: string, languageVersion: ScriptTarget, onError ?: (message: string) => void) {
+            // Return existing SourceFile object if one is available
+            if (cachedSourceFiles) {

```

Comment:
@mhegazy So, it's definitely simpler to do it the way you suggest, but one issue is that we'll keep alive old ASTs even when we know they've been modified (because we can't make the old program forget the AST). It's probably not a big deal (rare for more than one file to have been modified) and I still like the simplicity of having only the program object manage the cache and lookup logic.